doc/shared/nvectors/NVector_Kokkos.rst

```diff
@@ -0,0 +1,207 @@
+
+.. _NVectors.Kokkos:
+
+The NVECTOR_KOKKOS Module
+=========================
+
+.. versionadded:: 6.4.0
+
+The NVECTOR_KOKKOS :c:type:`N_Vector` implementation provides a vector data
+structure using Kokkos :cite:p:`edwards2014kokkos,trott2022kokkos` to support a
+variety of backends including serial, OpenMP, CUDA, HIP, and SYCL. Since Kokkos is
+a modern C++ library, the module is also written in modern C++ (it requires
+C++14) as a header only library. To utilize this ``N_Vector`` users will need to
+include ``nvector/nvector_kokkos.hpp``. More instructions on building SUNDIALS
+with Kokkos enabled are given in :numref:`Installation.CMake.ExternalLibraries`.
+For instructions on building and using Kokkos, refer to the `Kokkos
+<https://kokkos.github.io/kokkos-core-wiki/index.html>`_ documentation.
+
+.. _NVectors.Kokkos.Usage:
+
+Using NVECTOR_KOKKOS
+--------------------
+
+The NVECTOR_KOKKOS module is defined by the ``Vector`` templated
+class in the ``sundials::kokkos`` namespace:
+
+.. code-block:: cpp
+
+   template<class ExecutionSpace = Kokkos::DefaultExecutionSpace,
+            class MemorySpace = class ExecutionSpace::memory_space>
+      class Vector : public sundials::impl::BaseNVector,
+                     public sundials::ConvertibleTo<N_Vector>
+
+To use the NVECTOR_KOKKOS module, we construct an instance of the ``Vector`` class e.g.,
+
+.. code-block:: cpp
+
+   // Vector with extent length using the default execution space
+   sundials::kokkos::Vector<> x{length, sunctx};
+
+   // Vector with extent length using the Cuda execution space
+   sundials::kokkos::Vector<Kokkos::Cuda> x{length, sunctx};
+
+   // Vector based on an existing Kokkos::View
+   Kokkos::View<> view{"a view", length};
+   sundials::kokkos::Vector<> x{view, sunctx};
+
+   // Vector based on an existing Kokkos::View for device and host
+   Kokkos::View<Kokkos::Cuda> device_view{"a view", length};
+   Kokkos::View<Kokkos::HostMirror> host_view{Kokkos::create_mirror_view(device_view)};
+   sundials::kokkos::Vector<> x{device_views, host_view, sunctx};
+
+
+Instances of the ``Vector`` class are implicitly or explicitly (using the
+:cpp:func:`Vector::Convert` method) are convertible to a :c:type:`N_Vector`
+e.g.,
+
+.. code-block:: cpp
+
+   sundials::kokkos::Vector<> x{length, sunctx};
+   N_Vector x2 = x;           // implicit conversion to N_Vector
+   N_Vector x3 = x.Convert(); // explicit conversion to N_Vector
+
+No further interaction with a ``Vector`` is required from this point, and
+it is possible to use the :c:type:`N_Vector` API to operate on ``x2`` or ``x3``.
+
+.. warning::
+
+   :c:func:`N_VDestroy` should never be called on a ``N_Vector`` that was
+   created via conversion from a ``sundials::kokkos::Vector``. Doing so may
+   result in a double free.
+
+The underlying ``Vector`` can be extracted from a ``N_Vector`` using
+:cpp:func:`GetVec` e.g.,
+
+.. code-block:: cpp
+
+   auto x_vec = GetVec<>(x_nvector);
```

Comment:
```suggestion
   auto x_vec = GetVec<>(x3);
```
Given that just a few lines above we created `N_Vector` objects `x2` and `x3` that clearly contained a Kokkos vector, I'd like for us to use one of those here, instead of a generic name `x_nvector` where readers must assume its type.